File: electron/providers/tools/glob.ts
/**
 * Glob Tool - Find files by pattern
 */

import * as fs from "fs";
import * as path from "path";

export interface GlobInput {
  pattern: string;
  path?: string;
}

export interface GlobOutput {
  files: string[];
  count: number;
}

/**
 * Simple glob implementation without external dependencies
 * Supports basic patterns: *, **, ?
 */
export function globTool(input: GlobInput, cwd: string): GlobOutput {
  const basePath = input.path
    ? path.isAbsolute(input.path)
      ? input.path
      : path.resolve(cwd, input.path)
    : cwd;

  const pattern = input.pattern;
  const files: string[] = [];

  // Convert glob pattern to regex
  const regexPattern = globToRegex(pattern);

  // Walk directory tree
  walkDir(basePath, basePath, regexPattern, files);

  // Sort by modification time (newest first)
  files.sort((a, b) => {
    try {
      const statA = fs.statSync(a);
      const statB = fs.statSync(b);
      return statB.mtimeMs - statA.mtimeMs;
    } catch {
      return 0;
    }
  });

  return {
    files,
    count: files.length,
  };
}

function globToRegex(pattern: string): RegExp {
  let regex = pattern
    // Escape special regex chars (except * and ?)
    .replace(/[.+^${}()|[\]\\]/g, "\\$&")
    // ** matches any path
    .replace(/\*\*/g, "{{DOUBLESTAR}}")
    // * matches anything except /
    .replace(/\*/g, "[^/\\\\]*")
    // ? matches single char
    .replace(/\?/g, ".")
    // Restore **
    .replace(/{{DOUBLESTAR}}/g, ".*");

  return new RegExp(`^${regex}$`);
}

function walkDir(
  dir: string,
  basePath: string,
  pattern: RegExp,
  results: string[],
  maxDepth = 20
): void {
  if (maxDepth <= 0) return;

  try {
    const entries = fs.readdirSync(dir, { withFileTypes: true });

    for (const entry of entries) {
      // Skip hidden files and common ignored directories
      if (entry.name.startsWith(".") ||
          entry.name === "node_modules" ||
          entry.name === "__pycache__") {
        continue;
      }

      const fullPath = path.join(dir, entry.name);
      const relativePath = path.relative(basePath, fullPath).replace(/\\/g, "/");

      if (entry.isDirectory()) {
        walkDir(fullPath, basePath, pattern, results, maxDepth - 1);
      } else if (entry.isFile()) {
        if (pattern.test(relativePath) || pattern.test(entry.name)) {
          results.push(fullPath);
        }
      }
    }
  } catch {
    // Ignore permission errors
  }
}

File: src-tauri/src/providers/tools/glob.rs
use std::path::Path;

use anyhow::Result;

/// Find files matching a glob pattern.
///
/// * `pattern` - The glob pattern (e.g., `"**/*.ts"`).
/// * `path` - Optional base directory to search in (absolute or relative to `cwd`).
/// * `cwd` - Current working directory for resolving relative paths.
///
/// Returns a sorted `Vec<String>` of matching file paths (newest first).
pub fn glob_tool(
    pattern: &str,
    path: Option<&str>,
    cwd: &str,
) -> Result<Vec<String>> {
    let base_path = match path {
        Some(p) if Path::new(p).is_absolute() => p.to_string(),
        Some(p) => Path::new(cwd).join(p).to_string_lossy().to_string(),
        None => cwd.to_string(),
    };

    let walker = globwalk::GlobWalkerBuilder::from_patterns(&base_path, &[pattern])
        .max_depth(20)
        .follow_links(false)
        .build()?;

    let mut entries: Vec<(String, std::time::SystemTime)> = Vec::new();

    for entry in walker {
        let entry = match entry {
            Ok(e) => e,
            Err(_) => continue,
        };

        let path = entry.path();

        // Skip hidden files/dirs, node_modules, __pycache__
        let skip = path.components().any(|c| {
            let s = c.as_os_str().to_string_lossy();
            s.starts_with('.') || s == "node_modules" || s == "__pycache__"
        });
        if skip {
            continue;
        }

        if path.is_file() {
            let mtime = path
                .metadata()
                .and_then(|m| m.modified())
                .unwrap_or(std::time::UNIX_EPOCH);
            entries.push((path.to_string_lossy().to_string(), mtime));
        }
    }

    // Sort by modification time (newest first)
    entries.sort_by(|a, b| b.1.cmp(&a.1));

    let files: Vec<String> = entries.into_iter().map(|(path, _)| path).collect();
    Ok(files)
}
